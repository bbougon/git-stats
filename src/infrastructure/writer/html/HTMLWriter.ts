import * as fs from "fs";
import { intlFormat } from "date-fns";
import { __dirname } from "../FilePathConstant.js";
import * as pug from "pug";
import * as path from "path";
import { StatisticsAggregate } from "../../../statistics/GitStatistics.js";
import { Writer } from "../../../../index.js";
import { progressBar } from "../../progress-bar/ProgressBar.js";
import { Type } from "../../progress-bar/Type.js";
import { openBrowser } from "../OpenBrowser.js";
import { MergeEventsContentBuilder } from "./MergeEventsContentBuilder.js";
import { IssuesEventsContentBuilder } from "./IssuesEventsContentBuilder.js";

class HTMLContentBuilder {
  constructor(private readonly stats: StatisticsAggregate) {}

  build = (): string => {
    const humanizeDate = (date: Date): string => {
      return intlFormat(date, {
        weekday: "long",
        year: "numeric",
        month: "long",
        day: "numeric",
      });
    };
    const stringify = (obj: string): string => {
      return JSON.stringify(obj)
        .replace(/\u2028/g, "\\u2028")
        .replace(/\u2029/g, "\\u2029")
        .replace(/</g, "\\u003C")
        .replace(/>/g, "\\u003E")
        .replace(/\//g, "\\u002F");
    };
    const mergeEvents = new MergeEventsContentBuilder(this.stats).build();
    const issuesEvents = new IssuesEventsContentBuilder(this.stats).build();
    const start = humanizeDate(this.stats.mergeEvents.period.start);
    const end = humanizeDate(this.stats.mergeEvents.period.end);
    const templateFilePath = path.resolve(__dirname, "../../../templates/template.pug");
    const fn = pug.compileFile(templateFilePath, { pretty: true });
    return fn({
      stringify,
      period: { start, end },
      stats: {
        merge: { ...mergeEvents },
        issues: { ...issuesEvents },
      },
    });
  };
}

export class HTMLWriter implements Writer {
  private readonly _filePath: string;

  constructor(filePath: string) {
    this._filePath = filePath;
  }

  @progressBar(Type.Generate_HTML)
  write(stats: StatisticsAggregate): void {
    try {
      const reportFilePath = `${this._filePath}/report`;
      if (!fs.existsSync(reportFilePath)) {
        fs.mkdirSync(reportFilePath);
      }
      fs.writeFileSync(`${reportFilePath}/index.html`, new HTMLContentBuilder(stats).build());
      const cssFilePath = path.resolve(__dirname, "../../../style.css");
      fs.copyFileSync(cssFilePath, `${reportFilePath}/style.css`);
      openBrowser(`${reportFilePath}/index.html`);
    } catch (e) {
      console.log(e);
    }
  }
}
