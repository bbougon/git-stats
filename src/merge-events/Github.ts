import { RequestParameters } from "../../index.js";

export type PullRequestsStatsParameter = RequestParameters & {
  repo: string;
  owner: string;
};
