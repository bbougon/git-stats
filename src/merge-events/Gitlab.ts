import { RequestParameters } from "../../index.js";

export type MergeRequestsStatsParameters = RequestParameters & {
  projectId: number;
};
