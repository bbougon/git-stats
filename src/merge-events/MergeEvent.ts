import { compareAsc, differenceInHours, getMonth, getWeek, intervalToDuration } from "date-fns";
import { Repository } from "../Repository.js";
import { RequestParameters } from "../../index.js";
import moment from "moment";

export type MergeEvent = {
  project: number | string | undefined;
  id: number;
  createdAt: Date;
  mergedAt: Date | null;
  closedAt: Date | null;
};

export interface MergeEventRepository extends Repository<MergeEvent> {
  getMergeEventsForPeriod(requestParameters: RequestParameters): Promise<MergeEvent[]>;
}

type MergeEventsStatisticsResult = {
  average: {
    months: number;
    days: number;
    hours: number;
    minutes: number;
    seconds: number;
  };
  total: {
    merged: number;
    closed: number;
    opened: number;
    all: number;
  };
};

type Period = { start: Date; end: Date };

export class GitStatistics {
  constructor(private readonly mergeEvents: MergeEvent[], public readonly period: Period) {}

  public sortedMergeEvents(): MergeEvent[] {
    return this.mergeEvents
      .sort((mr, mrToCompare) => compareAsc(mr.createdAt, mrToCompare.createdAt))
      .sort((mr, mrToCompare) => compareAsc(mr.mergedAt, mrToCompare.mergedAt));
  }

  result = (): MergeEventsStatisticsResult => {
    const mergedMergeRequests = this.mergeEvents.filter((mr) => mr.mergedAt !== null);
    const closedMergeRequests = this.mergeEvents.filter((mr) => mr.closedAt !== null);
    const openedMergeRequests = this.mergeEvents.filter((mr) => mr.mergedAt === null && mr.closedAt == null);
    const hoursSpent = mergedMergeRequests.reduce(
      (accumulator, currentValue) => accumulator + differenceInHours(currentValue.mergedAt, currentValue.createdAt),
      0
    );
    const duration = moment.duration(hoursSpent / mergedMergeRequests.length, "hours");
    return {
      average: {
        months: parseInt(duration.months().toFixed()),
        days: parseInt(duration.days().toFixed()),
        hours: parseInt(duration.hours().toFixed()),
        minutes: parseInt(duration.minutes().toFixed()),
        seconds: parseInt(duration.seconds().toFixed()),
      },
      total: {
        merged: mergedMergeRequests.length,
        closed: closedMergeRequests.length,
        opened: openedMergeRequests.length,
        all: this.mergeEvents.length,
      },
    };
  };
}

export const mergeEventsStatistics = (
  requestParameter: RequestParameters,
  repository: MergeEventRepository
): Promise<GitStatistics> => {
  return repository.getMergeEventsForPeriod(requestParameter).then((mergeEvents) => {
    return new GitStatistics(mergeEvents, { end: requestParameter.toDate, start: requestParameter.fromDate });
  });
};

type Unit = string | "Week" | "Month";
export class Dimension {
  private constructor(public readonly unit: Unit, public readonly index: number, public mr: number) {}

  static create(unit: Unit, index: number) {
    return new Dimension(unit, index, 1);
  }

  increase() {
    this.mr = this.mr + 1;
  }

  static empty(unit: Unit, index: number) {
    return new Dimension(unit, index, 0);
  }
}

function fillEmptyPeriodsAndSortChronologically(
  stats: [number, [number, Dimension][]][],
  unit: string,
  period: Period
) {
  const result: Dimension[] = [];

  const fillPeriodTail = (stat: [number, [number, Dimension][]], dimensions: Dimension[]): void => {
    let periodEndIndex: number | undefined = undefined;
    if (stat[0] === period.end.getFullYear()) {
      periodEndIndex = unit === "Week" ? getWeek(period.end) : getMonth(period.end);
    }
    if (
      periodEndIndex !== undefined &&
      dimensions.find((dimension) => dimension.index === periodEndIndex) === undefined
    ) {
      const dimensionsLength = Math.max(...dimensions.map((dimension) => dimension.index));
      for (let i = 0; i < periodEndIndex - dimensionsLength; i++) {
        dimensions.push(Dimension.empty(unit, periodEndIndex - i));
      }
    }
  };

  const fillPeriodInterval = (stat: [number, [number, Dimension][]], dimensions: Dimension[]): void => {
    const periods = dimensions.map((val) => val.index);
    const periodStartIndex =
      stat[0] === period.start.getFullYear() ? (unit === "Week" ? getWeek(period.start) : getMonth(period.start)) : 0;
    for (let i = 0; i < periodStartIndex + periods.length; i++) {
      if (periodStartIndex <= i) {
        let j = i;
        while (j < periods[i - periodStartIndex]) {
          if (dimensions.find((dimension) => dimension.index === j) === undefined) {
            dimensions.push(Dimension.empty(unit, j));
          }
          j++;
        }
      }
    }
  };

  stats.forEach((stat) => {
    const flattenDimensions = stat[1]
      .flatMap((val) => val[1])
      .flatMap((val) => val)
      .sort((dimension, nextDimension) => (dimension.index > nextDimension.index ? 1 : -1));
    fillPeriodInterval(stat, flattenDimensions);
    fillPeriodTail(stat, flattenDimensions);
    result.push(...flattenDimensions.sort((stat, nextStat) => (stat.index > nextStat.index ? 1 : -1)));
  });
  return result;
}

export const mergeEventsByPeriod = (mergeEventsStatistics: GitStatistics): Dimension[] => {
  const stats: [number, [number, Dimension][]][] = [];
  const duration = intervalToDuration({
    start: mergeEventsStatistics.period.start,
    end: mergeEventsStatistics.period.end,
  });
  const moreThan2Months = duration.months > 1 && duration.months + duration.days > 2;
  const unit = moreThan2Months ? "Month" : "Week";
  mergeEventsStatistics
    .sortedMergeEvents()
    .filter((mr) => mr.mergedAt !== null)
    .forEach((mr) => {
      const year = mr.mergedAt.getFullYear();
      const dimension = Dimension.create(unit, moreThan2Months ? getMonth(mr.mergedAt) : getWeek(mr.mergedAt));
      if (stats.length === 0) {
        stats.push([year, [[dimension.index, dimension]]]);
      } else {
        const yearStats = stats.filter((stat) => stat[0] === year);
        if (yearStats.length === 0) {
          stats.push([year, [[dimension.index, dimension]]]);
        } else {
          const periodStats = yearStats[0][1].filter((stat) => stat[0] === dimension.index);
          if (periodStats.length === 0) {
            yearStats[0][1].push([dimension.index, dimension]);
          } else {
            const dimension = periodStats[0][1];
            dimension.increase();
          }
        }
      }
    });
  return fillEmptyPeriodsAndSortChronologically(stats, unit, mergeEventsStatistics.period);
};
