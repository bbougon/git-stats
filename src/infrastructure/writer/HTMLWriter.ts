import { Writer } from "../../../index.js";
import * as fs from "fs";
import { intlFormat } from "date-fns";
import { openBrowser } from "./OpenBrowser.js";
import * as pug from "pug";
import * as path from "path";
import { __dirname } from "./FilePathConstant.js";
import { mergedEventsStatisticByPeriod, StatisticsAggregate } from "../../statistics/GitStatistics.js";
import { MergeEventStatistics, MergeEvent } from "../../statistics/merge-events/MergeEvent.js";
import { progressBar } from "../progress-bar/ProgressBar.js";
import { Title } from "../progress-bar/Title.js";
import { HUMAN_READABLE_MONTHS } from "./HumanReadableLabels.js";
import moment from "moment/moment.js";
import Duration from "../../statistics/Duration.js";
import { CumulativeStatistics } from "../../statistics/CumulativeStatistics.js";

class HTMLContentBuilder {
  constructor(private readonly stats: StatisticsAggregate) {}

  build = (): string => {
    const humanizeDate = (date: Date): string => {
      return intlFormat(date, {
        weekday: "long",
        year: "numeric",
        month: "long",
        day: "numeric",
      });
    };
    const stringify = (obj: string): string => {
      return JSON.stringify(obj)
        .replace(/\u2028/g, "\\u2028")
        .replace(/\u2029/g, "\\u2029")
        .replace(/</g, "\\u003C")
        .replace(/>/g, "\\u003E")
        .replace(/\//g, "\\u002F");
    };
    const {
      mergedEventMonthsLabels,
      mergedEventsMonthsData,
      mergedEventsMonthsAverageTimeData,
      mergedEventsMonthsMedianTimeData,
      mergedEventsWeeksLabels,
      mergedEventsWeeksData,
      mergedEventsWeeksAverageData,
      mergedEventsWeeksMedianData,
    } = this.mergedEventsStatistics();

    const {
      cumulativeMonthsLabels,
      cumulativeOpenedMonthsData,
      cumulativeClosedMonthsData,
      cumulativeTrendMonthsData,
      cumulativeWeeksLabels,
      cumulativeOpenedWeeksData,
      cumulativeClosedWeeksData,
      cumulativeTrendWeeksData,
    } = this.cumulativeStatistics();

    const aggregatedStats = this.stats.mergedEvents.result();
    const start = humanizeDate(this.stats.mergedEvents.period.start);
    const end = humanizeDate(this.stats.mergedEvents.period.end);
    const templateFilePath = path.resolve(__dirname, "../../../templates/template.pug");
    const fn = pug.compileFile(templateFilePath, { pretty: true });
    return fn({
      stringify,
      period: { start, end },
      stats: {
        mr: {
          months: {
            data: mergedEventsMonthsData,
            labels: mergedEventMonthsLabels,
            average: mergedEventsMonthsAverageTimeData,
            median: mergedEventsMonthsMedianTimeData,
          },
          weeks: {
            data: mergedEventsWeeksData,
            labels: mergedEventsWeeksLabels,
            average: mergedEventsWeeksAverageData,
            median: mergedEventsWeeksMedianData,
          },
        },
        cumulative: {
          months: {
            openedData: cumulativeOpenedMonthsData,
            closedData: cumulativeClosedMonthsData,
            trendData: cumulativeTrendMonthsData,
            labels: cumulativeMonthsLabels,
          },
          weeks: {
            openedData: cumulativeOpenedWeeksData,
            closedData: cumulativeClosedWeeksData,
            trendData: cumulativeTrendWeeksData,
            labels: cumulativeWeeksLabels,
          },
        },
        ...aggregatedStats,
      },
    });
  };

  private mergedEventsStatistics() {
    const mergedEventsStatistics = mergedEventsStatisticByPeriod(
      this.stats.mergedEvents as MergeEventStatistics,
      (mr: MergeEvent) => ({
        end: mr.mergedAt,
        start: mr.createdAt,
      })
    );
    const mergedEventMonthsLabels: string[] = [];
    const mergedEventsMonthsData: number[] = [];
    const mergedEventsMonthsAverageTimeData: number[] = [];
    const mergedEventsMonthsMedianTimeData: number[] = [];
    const mergedEventsWeeksLabels: string[] = [];
    const mergedEventsWeeksData: number[] = [];
    const mergedEventsWeeksAverageData: number[] = [];
    const mergedEventsWeeksMedianData: number[] = [];

    function durationAsHours(duration: Duration) {
      return moment
        .duration(duration.months, "months")
        .add(duration.days, "days")
        .add(duration.hours, "hours")
        .add(duration.minutes, "minutes")
        .add(duration.seconds, "seconds")
        .asHours();
    }

    mergedEventsStatistics.forEach((stat) => {
      stat.forEach((period) => {
        Object.entries(period).forEach(([unit, flows]) => {
          if (unit === "Month") {
            flows.forEach((flow) => {
              mergedEventMonthsLabels.push(HUMAN_READABLE_MONTHS[flow.index]);
              mergedEventsMonthsData.push(flow.total());
              mergedEventsMonthsAverageTimeData.push(durationAsHours(flow.average()));
              mergedEventsMonthsMedianTimeData.push(durationAsHours(flow.median()));
            });
          }
          if (unit === "Week") {
            flows.forEach((flow) => {
              mergedEventsWeeksLabels.push(`Week ${flow.index}`);
              mergedEventsWeeksData.push(flow.total());
              mergedEventsWeeksAverageData.push(durationAsHours(flow.average()));
              mergedEventsWeeksMedianData.push(durationAsHours(flow.median()));
            });
          }
        });
      });
    });
    return {
      mergedEventMonthsLabels,
      mergedEventsMonthsData,
      mergedEventsMonthsAverageTimeData,
      mergedEventsMonthsMedianTimeData,
      mergedEventsWeeksLabels,
      mergedEventsWeeksData,
      mergedEventsWeeksAverageData,
      mergedEventsWeeksMedianData,
    };
  }

  private cumulativeStatistics() {
    const cumulativeMonthsLabels: string[] = [];
    const cumulativeOpenedMonthsData: number[] = [];
    const cumulativeClosedMonthsData: number[] = [];
    const cumulativeTrendMonthsData: number[] = [];
    const cumulativeWeeksLabels: string[] = [];
    const cumulativeOpenedWeeksData: number[] = [];
    const cumulativeClosedWeeksData: number[] = [];
    const cumulativeTrendWeeksData: number[] = [];
    const cumulativeStatistics = (this.stats.cumulativeStatistics as CumulativeStatistics).result().cumulativeResults;

    cumulativeStatistics.get("Month").forEach((cumulativeStatistics) => {
      cumulativeMonthsLabels.push(HUMAN_READABLE_MONTHS[cumulativeStatistics.index]);
      cumulativeOpenedMonthsData.push(cumulativeStatistics.opened);
      cumulativeClosedMonthsData.push(cumulativeStatistics.closed);
      cumulativeTrendMonthsData.push(cumulativeStatistics.trend);
    });

    cumulativeStatistics.get("Week").forEach((cumulativeStatistics) => {
      cumulativeWeeksLabels.push(`Week ${cumulativeStatistics.index}`);
      cumulativeOpenedWeeksData.push(cumulativeStatistics.opened);
      cumulativeClosedWeeksData.push(cumulativeStatistics.closed);
      cumulativeTrendWeeksData.push(cumulativeStatistics.trend);
    });

    return {
      cumulativeMonthsLabels,
      cumulativeOpenedMonthsData,
      cumulativeClosedMonthsData,
      cumulativeTrendMonthsData,
      cumulativeWeeksLabels,
      cumulativeOpenedWeeksData,
      cumulativeClosedWeeksData,
      cumulativeTrendWeeksData,
    };
  }
}

export class HTMLWriter implements Writer {
  private _filePath: string;

  constructor(filePath: string) {
    this._filePath = filePath;
  }

  @progressBar(Title.Generate_HTML)
  write(stats: StatisticsAggregate): void {
    try {
      const reportFilePath = `${this._filePath}/report`;
      if (!fs.existsSync(reportFilePath)) {
        fs.mkdirSync(reportFilePath);
      }
      fs.writeFileSync(`${reportFilePath}/index.html`, new HTMLContentBuilder(stats).build());
      const cssFilePath = path.resolve(__dirname, "../../../style.css");
      fs.copyFileSync(cssFilePath, `${reportFilePath}/style.css`);
      openBrowser(`${reportFilePath}/index.html`);
    } catch (e) {
      console.log(e);
    }
  }
}
