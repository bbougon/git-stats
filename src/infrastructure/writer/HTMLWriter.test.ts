import { jest } from "@jest/globals";
import { HTMLWriter } from "./HTMLWriter";
import { mkdtemp } from "node:fs/promises";
import * as path from "path";
import * as os from "os";
import * as fs from "fs";
import { MergeEventBuilderForMR } from "../../__tests__/builder";
import { parseISO } from "date-fns";
import { MergedEventStatistics } from "../../statistics/merge-events/MergeEvent";

jest.mock("./FilePathConstant", () => ({
  __dirname: "src/infrastructure/writer/",
}));

jest.mock("./OpenBrowser", () => ({
  openBrowser: jest.fn(),
}));

describe("HTML writer", () => {
  test("should generate an HTML report file for merged requests", async () => {
    const firstMergeRequest = new MergeEventBuilderForMR(1)
      .createdAt(parseISO("2022-02-11T12:37:22"))
      .mergedAt(parseISO("2022-02-14T11:53:17"))
      .build();
    const secondMergeRequest = new MergeEventBuilderForMR(1)
      .createdAt(parseISO("2022-02-12T13:22:54"))
      .mergedAt(parseISO("2022-02-12T18:17:32"))
      .build();
    const thirdMergeRequest = new MergeEventBuilderForMR(1)
      .createdAt(parseISO("2022-02-13T09:17:34"))
      .mergedAt(parseISO("2022-02-16T16:44:22"))
      .build();
    const tempDirectory = await mkdtemp(path.join(os.tmpdir(), "report-"));
    const fromDate = parseISO("2022-02-11T00:00:00");
    const toDate = parseISO("2022-02-17T00:00:00");

    new HTMLWriter(tempDirectory).write({
      mergedEvents: new MergedEventStatistics([firstMergeRequest, secondMergeRequest, thirdMergeRequest], {
        end: toDate,
        start: fromDate,
      }),
    });

    expect(fs.readFileSync(`${tempDirectory}/report/index.html`, "utf8")).toMatchSnapshot();
  });

  test("should generate an HTML report file for merged requests when some mr are not yet merged", async () => {
    const firstMergeRequest = new MergeEventBuilderForMR(1)
      .createdAt(parseISO("2022-02-11T12:37:22"))
      .mergedAt(parseISO("2022-02-14T11:53:17"))
      .build();
    const secondMergeRequest = new MergeEventBuilderForMR(1)
      .createdAt(parseISO("2022-02-12T13:22:54"))
      .notYetMerged()
      .build();
    const thirdMergeRequest = new MergeEventBuilderForMR(1)
      .createdAt(parseISO("2022-02-13T09:17:34"))
      .mergedAt(parseISO("2022-02-16T16:44:22"))
      .build();
    const tempDirectory = await mkdtemp(path.join(os.tmpdir(), "report-"));
    const fromDate = parseISO("2022-02-11T00:00:00");
    const toDate = parseISO("2022-02-17T00:00:00");

    new HTMLWriter(tempDirectory).write({
      mergedEvents: new MergedEventStatistics([firstMergeRequest, secondMergeRequest, thirdMergeRequest], {
        end: toDate,
        start: fromDate,
      }),
    });

    expect(fs.readFileSync(`${tempDirectory}/report/index.html`, "utf8")).toMatchSnapshot();
  });

  test("should generate an HTML report file with months period", async () => {
    const firstMergeRequest = new MergeEventBuilderForMR(1)
      .createdAt(parseISO("2022-02-11T12:37:22"))
      .mergedAt(parseISO("2022-02-14T11:53:17"))
      .build();
    const secondMergeRequest = new MergeEventBuilderForMR(1)
      .createdAt(parseISO("2022-01-10T13:22:54"))
      .mergedAt(parseISO("2022-01-27T13:22:54"))
      .build();
    const thirdMergeRequest = new MergeEventBuilderForMR(1)
      .createdAt(parseISO("2022-02-13T09:17:34"))
      .mergedAt(parseISO("2022-02-16T16:44:22"))
      .build();
    const tempDirectory = await mkdtemp(path.join(os.tmpdir(), "report-"));

    new HTMLWriter(tempDirectory).write({
      mergedEvents: new MergedEventStatistics([firstMergeRequest, secondMergeRequest, thirdMergeRequest], {
        end: parseISO("2022-03-02T00:00:00"),
        start: parseISO("2022-01-01T00:00:00"),
      }),
    });

    expect(fs.readFileSync(`${tempDirectory}/report/index.html`, "utf8")).toMatchSnapshot();
  });
});
