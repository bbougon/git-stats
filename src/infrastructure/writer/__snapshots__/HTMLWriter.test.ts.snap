// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML writer should generate an HTML report file for merged requests 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Report Friday, February 11, 2022 to Thursday, February 17, 2022</title>
    <link href="style.css" rel="stylesheet">
    <script src="https://cdn.jsdelivr.net/npm/chart.js"></script>
  </head>
  <body>
    <div>
      <h2>Merged events for period Friday, February 11, 2022 to Thursday, February 17, 2022</h2>
      <div class="container-sm text-left">
        <h3>Aggregated Statistics</h3>
        <div class="row">
          <div class="col-lg-10 text-secondary">
            <div class="row">
              <div class="col-sm-12 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title">0 months, 2 days, 3 hours, 20 minutes, 0 seconds</h5>
                    <p class="card-text"><em>Average time to merge</em></p>
                  </div>
                </div>
              </div>
            </div>
            <div class="row">
              <div class="col-sm-3 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title text-merged">3</h5>
                    <p class="card-text"><em># Merged</em></p>
                  </div>
                </div>
              </div>
              <div class="col-sm-3 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title text-opened">0</h5>
                    <p class="card-text"><em># Opened</em></p>
                  </div>
                </div>
              </div>
              <div class="col-sm-3 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title text-closed">0</h5>
                    <p class="card-text"><em># Closed</em></p>
                  </div>
                </div>
              </div>
              <div class="col-sm-3 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title text-total">3</h5>
                    <p class="card-text"><em># Overall</em></p>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="col-lg-2">
            <div class="row">
              <div>
                <canvas id="aggregated-statistics"></canvas>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="container-sm">
        <canvas id="merged-events"></canvas>
      </div>
    </div>
    <script type="text/javascript">
      const mergeRequestContext = document.getElementById('merged-events');
      const mergeRequestChartType = 'line'
      const mergeRequestChartDatasets = [{
          label: '# of Merge Requests merged',
          data: [1,2],
          borderWidth: 3,
          borderJoinStyle: "round",
          tension: 0.2
      }]
      const mergeRequestChartLabels = ["Week 7","Week 8"]
      
      const aggregatedMergeRequestContext = document.getElementById("aggregated-statistics")
      const aggregatedMergeRequestChartType = 'doughnut'
      const aggregatedMergeRequestChartLabels = ['Merged', 'Opened', 'Closed', 'All']
      const aggregatedMergeRequestChartDatasets =    [{
          label: 'Aggregated merge requests',
          data: [
              3,
              0,
              0,
              3
          ],
          hoverOffset: 4
      }]
      
      function renderChart(context, type, labels, datasets) {
          new Chart(context, {
              type: type,
              data: {
                  labels: labels,
                  datasets: datasets,
              },
              options: {}
          });
      }
      
      renderChart(mergeRequestContext, mergeRequestChartType, mergeRequestChartLabels, mergeRequestChartDatasets);
      renderChart(aggregatedMergeRequestContext, aggregatedMergeRequestChartType, aggregatedMergeRequestChartLabels, aggregatedMergeRequestChartDatasets);
    </script>
  </body>
</html>"
`;

exports[`HTML writer should generate an HTML report file for merged requests when some mr are not yet merged 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Report Friday, February 11, 2022 to Thursday, February 17, 2022</title>
    <link href="style.css" rel="stylesheet">
    <script src="https://cdn.jsdelivr.net/npm/chart.js"></script>
  </head>
  <body>
    <div>
      <h2>Merged events for period Friday, February 11, 2022 to Thursday, February 17, 2022</h2>
      <div class="container-sm text-left">
        <h3>Aggregated Statistics</h3>
        <div class="row">
          <div class="col-lg-10 text-secondary">
            <div class="row">
              <div class="col-sm-12 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title">0 months, 3 days, 3 hours, 0 minutes, 0 seconds</h5>
                    <p class="card-text"><em>Average time to merge</em></p>
                  </div>
                </div>
              </div>
            </div>
            <div class="row">
              <div class="col-sm-3 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title text-merged">2</h5>
                    <p class="card-text"><em># Merged</em></p>
                  </div>
                </div>
              </div>
              <div class="col-sm-3 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title text-opened">1</h5>
                    <p class="card-text"><em># Opened</em></p>
                  </div>
                </div>
              </div>
              <div class="col-sm-3 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title text-closed">0</h5>
                    <p class="card-text"><em># Closed</em></p>
                  </div>
                </div>
              </div>
              <div class="col-sm-3 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title text-total">3</h5>
                    <p class="card-text"><em># Overall</em></p>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="col-lg-2">
            <div class="row">
              <div>
                <canvas id="aggregated-statistics"></canvas>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="container-sm">
        <canvas id="merged-events"></canvas>
      </div>
    </div>
    <script type="text/javascript">
      const mergeRequestContext = document.getElementById('merged-events');
      const mergeRequestChartType = 'line'
      const mergeRequestChartDatasets = [{
          label: '# of Merge Requests merged',
          data: [0,2],
          borderWidth: 3,
          borderJoinStyle: "round",
          tension: 0.2
      }]
      const mergeRequestChartLabels = ["Week 7","Week 8"]
      
      const aggregatedMergeRequestContext = document.getElementById("aggregated-statistics")
      const aggregatedMergeRequestChartType = 'doughnut'
      const aggregatedMergeRequestChartLabels = ['Merged', 'Opened', 'Closed', 'All']
      const aggregatedMergeRequestChartDatasets =    [{
          label: 'Aggregated merge requests',
          data: [
              2,
              1,
              0,
              3
          ],
          hoverOffset: 4
      }]
      
      function renderChart(context, type, labels, datasets) {
          new Chart(context, {
              type: type,
              data: {
                  labels: labels,
                  datasets: datasets,
              },
              options: {}
          });
      }
      
      renderChart(mergeRequestContext, mergeRequestChartType, mergeRequestChartLabels, mergeRequestChartDatasets);
      renderChart(aggregatedMergeRequestContext, aggregatedMergeRequestChartType, aggregatedMergeRequestChartLabels, aggregatedMergeRequestChartDatasets);
    </script>
  </body>
</html>"
`;

exports[`HTML writer should generate an HTML report file with months period 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Report Saturday, January 1, 2022 to Wednesday, March 2, 2022</title>
    <link href="style.css" rel="stylesheet">
    <script src="https://cdn.jsdelivr.net/npm/chart.js"></script>
  </head>
  <body>
    <div>
      <h2>Merged events for period Saturday, January 1, 2022 to Wednesday, March 2, 2022</h2>
      <div class="container-sm text-left">
        <h3>Aggregated Statistics</h3>
        <div class="row">
          <div class="col-lg-10 text-secondary">
            <div class="row">
              <div class="col-sm-12 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title">0 months, 7 days, 18 hours, 0 minutes, 0 seconds</h5>
                    <p class="card-text"><em>Average time to merge</em></p>
                  </div>
                </div>
              </div>
            </div>
            <div class="row">
              <div class="col-sm-3 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title text-merged">3</h5>
                    <p class="card-text"><em># Merged</em></p>
                  </div>
                </div>
              </div>
              <div class="col-sm-3 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title text-opened">0</h5>
                    <p class="card-text"><em># Opened</em></p>
                  </div>
                </div>
              </div>
              <div class="col-sm-3 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title text-closed">0</h5>
                    <p class="card-text"><em># Closed</em></p>
                  </div>
                </div>
              </div>
              <div class="col-sm-3 mb-3 mb-sm-0">
                <div class="card mb-3">
                  <div class="card-body text-center">
                    <h5 class="card-title text-total">3</h5>
                    <p class="card-text"><em># Overall</em></p>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="col-lg-2">
            <div class="row">
              <div>
                <canvas id="aggregated-statistics"></canvas>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="container-sm">
        <canvas id="merged-events"></canvas>
      </div>
    </div>
    <script type="text/javascript">
      const mergeRequestContext = document.getElementById('merged-events');
      const mergeRequestChartType = 'line'
      const mergeRequestChartDatasets = [{
          label: '# of Merge Requests merged',
          data: [1,2,0],
          borderWidth: 3,
          borderJoinStyle: "round",
          tension: 0.2
      }]
      const mergeRequestChartLabels = ["Jan","Feb","Mar"]
      
      const aggregatedMergeRequestContext = document.getElementById("aggregated-statistics")
      const aggregatedMergeRequestChartType = 'doughnut'
      const aggregatedMergeRequestChartLabels = ['Merged', 'Opened', 'Closed', 'All']
      const aggregatedMergeRequestChartDatasets =    [{
          label: 'Aggregated merge requests',
          data: [
              3,
              0,
              0,
              3
          ],
          hoverOffset: 4
      }]
      
      function renderChart(context, type, labels, datasets) {
          new Chart(context, {
              type: type,
              data: {
                  labels: labels,
                  datasets: datasets,
              },
              options: {}
          });
      }
      
      renderChart(mergeRequestContext, mergeRequestChartType, mergeRequestChartLabels, mergeRequestChartDatasets);
      renderChart(aggregatedMergeRequestContext, aggregatedMergeRequestChartType, aggregatedMergeRequestChartLabels, aggregatedMergeRequestChartDatasets);
    </script>
  </body>
</html>"
`;
